### ISSUE TITLE
BulkMerge ignores DateTime entity model properties backed by SQL datetime columns

### ISSUE THREAD

**Issue by csnelldaw** [NorthernLight1/N.EntityFrameworkCore.Extensions#83]
Saving a list of existing entities using BulkMerge with default BulkMergeOptions will update string and boolean values but DateTime values are being ignored.

Example entity model (EF core code first):

    public partial class ApplicationEvent 
    {
        public ApplicationEvent()  {  }

        [Key]
        [Column("ID")]
        public Guid Id { get; set; }  // has default value in SQL column of newid()

        [Required]
        [StringLength(50)]
        public string CreatedBy { get; set; }

        [Column(TypeName = "datetime")]
        public DateTime DateCreated { get; set; } // has default value in SQL column of getdate()

        [Required]
        [StringLength(50)]
        public string EventType { get; set; }

        [Required]
        [StringLength(1000)]
        public string Subject { get; set; }

        [StringLength(4000)]
        public string Message { get; set; }
    }

Below is SQL Profiler output during BulkMerge execution using default BulkMergeOptions.  Notice absence of the datetime column DateCreated except in the final OUTPUT statement.

SELECT TOP 0 [ApplicationEvents].[ID],[ApplicationEvents].[CreatedBy],[ApplicationEvents].[EventType],[ApplicationEvents].[Message],[ApplicationEvents].[Subject],CAST( NULL AS INT) AS _be_xx_id INTO [dbo].[#tmp_be_xx_ApplicationEvents] FROM [dbo].[ApplicationEvents]

insert bulk [dbo].[#tmp_be_xx_ApplicationEvents] ([ID] UniqueIdentifier, [CreatedBy] NVarChar(50) COLLATE SQL_Latin1_General_CP1_CI_AS, [EventType] NVarChar(50) COLLATE SQL_Latin1_General_CP1_CI_AS, [Message] NVarChar(4000) COLLATE SQL_Latin1_General_CP1_CI_AS, [Subject] NVarChar(1000) COLLATE SQL_Latin1_General_CP1_CI_AS, [_be_xx_id] Int)

MERGE [ApplicationEvents] AS t USING [dbo].[#tmp_be_xx_ApplicationEvents] AS s ON t.ID=s.ID WHEN NOT MATCHED THEN INSERT ([CreatedBy],[EventType],[Message],[Subject]) VALUES ([CreatedBy],[EventType],[Message],[Subject]) WHEN MATCHED THEN UPDATE SET t.[CreatedBy]=s.[CreatedBy],t.[EventType]=s.[EventType],t.[Message]=s.[Message],t.[Subject]=s.[Subject] OUTPUT $Action,[s].[_be_xx_id],[inserted].[ID],[inserted].[DateCreated];

Attempts to set the DateCreated to a different value while investigating this issue failed to update the database.  DateCreated is omitted from the select/insert/merge statements above (other than as OUTPUT) as if the column doesn't exist.  Another model with DateCreated and DateModified datetime columns exhibited the same behavior for both datetime columns.  DateModified is flagged as a concurrency token, but DateCreated is not.

Using EF Core 8.0.8 with SQL Server 2019 and N.EntityFrameworkCore.Extensions 8.0.0.11.

**Comment by NorthernLight1** [comment]
[csnelldaw](https://github.com/csnelldaw),

Can I see what entity options you are using to set CreatedDate?

**Comment by csnelldaw** [comment]
Thanks for the quick response.  The dates are being set manually prior to
bulk merge (I was testing N.EFCore.Ext in a component library I ported from
.net framework that uses N.EF.Ext).  Here are the snippets of code:

In a service class, a simple bulk retrieve, modify, and save attempt to
test.  With a different entity ('lookup item') class, I modified
DateModified along with a boolean field and string field.  Only the boolean
and string value changes were persisted:

            var items =
Repository.FindAll<ApplicationEvent>().Take(10).ToArray();
            items[0].DateCreated = new DateTime(2022, 1, 1);
            var x = Repository.SaveListAsBulk(items, state.AuditName);

In Repository.cs, SaveListAsBulk method

            BulkMergeResult<T> mergeResult =
dbContext.BulkMergeAudited(models); // models = list of ApplicationEvents

In db context base class that inherits from DbContext, BulkMergeAudited()
method:

           // this line normally sets DateCreated (new models only) and, if
present, DateModified, prior to saving; it's been commented out while
investigating this issue.
           // otherwise, the EntityHelper.SetAuditPropertiesFor method is
working correctly for conventional EF Core inserts and updates
           // models.ForEach(m => EntityHelper.SetAuditPropertiesFor(m,
m.IsNew ? EntityState.Added : EntityState.Modified, AuditName,
DisableAuditUpdate));

         return this.BulkMerge(models); // call to N.EFCore.Ext method.
The 1/1/22 date survives up to the point of execution but reverts back
after BulkMerge executes

Let me know if you have any questions.

Thanks,

Chuck

Chuck Snell

Data AnyWare, Inc.
Ph: 919-650-4650
data-anyware.com

*CONFIDENTIAL*: This email and any files transmitted with it are
confidential and intended solely for the use of the parties to whom this
email is addressed. If you have received this email in error, please notify
us immediately by replying to this email or contacting us at 919-932-1955,
and destroy all copies of this email and attached files. If you are not the
intended recipient you are hereby notified that disclosing, copying,
distributing or taking any action in reliance on the contents of this
information is strictly prohibited.

**Comment by NorthernLight1** [comment]
Does CreateDate and ModifiedDate use default and\or computed values? I believe that is the case based on your first response, but I would just like you to confirm it.

**Comment by csnelldaw** [comment]
They both use default values of getdate() in the column definitions. Values
are typically provided when records are created via code, but often times
records get generated/updated by sql scripts, which is why the defaults
exist.


Chuck Snell
Data AnyWare, Inc.
Ph: 919-650-4650
data-anyware.com

*CONFIDENTIAL*: This email and any files transmitted with it are
confidential and intended solely for the use of the parties to whom this
email is addressed. If you have received this email in error, please notify
us immediately by replying to this email or contacting us at 919-932-1955,
and destroy all copies of this email and attached files. If you are not the
intended recipient you are hereby notified that disclosing, copying,
distributing or taking any action in reliance on the contents of this
information is strictly prohibited.

**Comment by NorthernLight1** [comment]
Thanks, I will take a look it.

**Comment by csnelldaw** [comment]
After some digging I figured out why the date columns were not getting updated.  The column list generated in TableMapping.GetColumnNames requires column properties to have ValueGenerated == ValueGenerated.Never, but EF Core Power Tools (using Handlebars templating for reverse engineering the db context) adds .HasDefaultValueSql("(getdate())") to the entity DateCreated/DateModified properties per the default values in the SQL table definitions.  Same with some non-nullable bit columns having a default of 0.  

The BulkMerge behavior for default values makes sense, so I've adapted our db context to set ValueGenerated to ValueGenerated.Never during model creation in the context since the app controls the values (except PKs) anyway. The default values are more of a convenience for script-generated inserts.   BulkMerge is working as expected with this adjustment to our code.
